### PR TITLE
polyfill `Worker` in chrome runtime

### PR DESCRIPTION
**Pull request by turbocrime**
references https://github.com/penumbra-zone/web/issues/1795

this PR creates a package implementing a polyfill of the DOM `Worker` interface for chrome extension workers, and applies that polyfill to consumed packages from penumbra-zone.

pending https://github.com/penumbra-zone/web/pull/1769

### CODE REVIEW

**Comment by turbocrime** [comment]
managed to discover some lies within the chrome runtime api about the offscreen lifecycle, so i have to work around that, but this is nearly ready for review.

File: apps/extension/public/manifest.json [hunk 1]
```diff
@@ -32,7 +32,8 @@
     "open_in_tab": true
   },
   "background": {
-    "service_worker": "service-worker.js"
+    "service_worker": "service-worker.js",
+    "type": "module"
   },
```
Comment (turbocrime) [inline apps/extension/public/manifest.json:37]: this change is necessary due to some webpack reconfiguration.

File: apps/extension/src/service-worker.ts [hunk 2]
```diff
@@ -43,6 +43,15 @@ import { walletIdCtx } from '@penumbra-zone/services/ctx/wallet-id';
...
 
+import { OffscreenWorker } from '@repo/chrome-offscreen-worker/worker';
+
+globalThis.Worker = OffscreenWorker;
+OffscreenWorker.configure(
+  chrome.runtime.getURL(
+    'offscreen.html',
+  ) as `chrome-extension://${typeof chrome.runtime.id}/${'offscreen.html'}`,
+);
+
```
Comment (turbocrime) [inline apps/extension/src/service-worker.ts:54]: the offscreen worker polyfill is applied to the global here. any packages bundled into this entry calling `new Worker` will actually be calling `new OffscreenWorker`

File: package.json [hunk 3]
```diff
@@ -31,6 +31,7 @@
     "@bufbuild/protobuf": "^1.10.0",
     "@connectrpc/connect": "^1.4.0",
     "@connectrpc/connect-web": "^1.4.0",
+    "@tsconfig/strictest": "^2.0.5",
```
Comment (turbocrime) [inline package.json:34]: https://github.com/prax-wallet/web/issues/103

File: packages/chrome-offscreen-worker/src/control.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
+/// <reference types="chrome" />
+
+import { OffscreenWorkerPort } from './messages/worker-event.js';
+import { OffscreenRootPort, WorkerConstructorParamsPrimitive } from './messages/root-control.js';
+
+const attempts: string[] = [];
```
Comment (turbocrime) [inline packages/chrome-offscreen-worker/src/control.ts:6]: todo: remove debug

```suggestion
```

File: packages/chrome-offscreen-worker/src/control.ts [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+   * This class manages the offscreen document and worker connections. Your
+   * offscreen document should use the `entry` export of this package, a
+   * side-effectful module which attaches listeners and manages the actual
+   * workers.
+   *
+   * @param offscreenUrl a fully-qualified URL to your offscreen document, as output by `chrome.runtime.getURL`
+   * @param timeout wait for the offscreen document to connect to this controller, upon first worker init (10s default)
+   */
+  constructor(
+    offscreenUrl: `chrome-extension://${string}/${string}`,
```
Comment (turbocrime) [inline packages/chrome-offscreen-worker/src/control.ts:24]: todo: this is a little awkward for consumers

File: packages/chrome-offscreen-worker/src/control.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+      const offscreenContext = await hasOffscreen();
+      if (!offscreenContext?.documentUrl) {
+        console.log('createOffscreen !offscreenContext');
+        const offscreenInitUrl = new URL(`#${crypto.randomUUID()}`, this.offscreenPath);
+        console.log('initting', offscreenInitUrl.href);
+        await chrome.offscreen.createDocument({
+          url: offscreenInitUrl.href,
+          reasons: [chrome.offscreen.Reason.WORKERS],
+          justification:
+            'Manages workers for parallelizing heavy computation outside of the service worker thread',
```
Comment (turbocrime) [inline packages/chrome-offscreen-worker/src/control.ts:72]: todo: make `justification` part of the configuration step

File: packages/chrome-offscreen-worker/src/control.ts
Comment (turbocrime) [inline packages/chrome-offscreen-worker/src/control.ts:1]: plenty of `console.log` in here just close your eyes it's fine

File: packages/chrome-offscreen-worker/src/entry.ts
Comment (turbocrime) [inline packages/chrome-offscreen-worker/src/entry.ts:1]: imagine you cannot read the string `console.log`

File: packages/chrome-offscreen-worker/src/messages/root-control.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
+/* These messages appear only in the single offscreen root control channel. */
+
+export type WorkerConstructorParamsPrimitive =
+  Required<ConstructorParameters<typeof Worker>> extends [infer S, infer O]
+    ? [S extends string ? S : never, { [k in keyof O]: O[k] & string }]
+    : never;
```
Comment (turbocrime) [inline packages/chrome-offscreen-worker/src/messages/root-control.ts:6]: mostly this just excludes `URL` from the worker path url parameter, but it also types the `WorkerOptions` structure as primitives to indicate jsonifiability

File: packages/chrome-offscreen-worker/src/worker.ts [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    switch (evt.type) {
+      case 'error':
+      case 'message':
+      case 'messageerror':
+        void this.workerPort.then(port => port.postMessage(unknownMessage as OffscreenWorkerEvent));
+        return;
+      default:
+        throw new Error('Unknown event from parent', { cause: evt });
+    }
+  };
```
Comment (turbocrime) [inline packages/chrome-offscreen-worker/src/worker.ts:107]: this accepts pretty much anything. it probably needs some validation, but again, technically callers could dispatch pretty much anything.

File: packages/chrome-offscreen-worker/src/worker.ts [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+        case 'message':
+        case 'messageerror':
+          this.internalTarget.dispatchEvent(new MessageEvent(json.type, json.init));
+          return;
+        default:
+          //console.warn('Dispatching unknown event', json);
+          //this.internalTarget.dispatchEvent(new Event(json.type, json.init));
+          throw new Error('Unknown event in worker output', {
+            cause: { type: Boolean(json) && json.type, message: json, port: port.name },
+          });
```
Comment (turbocrime) [inline packages/chrome-offscreen-worker/src/worker.ts:90]: todo: technically, callers may dispatch any kind of event on a DOM worker instance, but it's not clear if all events propagate to the worker.

File: packages/chrome-offscreen-worker/src/worker.ts [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    );
+
+    void this.workerPort.then(
+      port => {
+        console.log(this.params[1].name, 'got chromePort');
+        port.onMessage.addListener(this.workerListener);
+
+        this.internalTarget.addEventListener('error', this.parentListener);
+        this.internalTarget.addEventListener('message', this.parentListener);
+        this.internalTarget.addEventListener('messageerror', this.parentListener);
```
Comment (turbocrime) [inline packages/chrome-offscreen-worker/src/worker.ts:66]: each item of `WorkerEventMap` is listened

File: packages/chrome-offscreen-worker/src/worker.ts [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+    void this.workerPort.then(port => port.disconnect());
+  };
+
+  postMessage: Worker['postMessage'] = (...args) => {
+    console.debug(this.params[1].name, 'postMessage', args);
+    // typescript doesn't handle the overloaded signature very well
+    const data: unknown = args[0];
+    const { transfer } = Array.isArray(args[1]) ? { transfer: args[1] } : { ...args[1] };
+
+    this.externalTarget.postMessage(data, { transfer });
```
Comment (turbocrime) [inline packages/chrome-offscreen-worker/src/worker.ts:120]: transferrable objects are accepted, but are likely to not serialize properly in `parentListener`

this doesn't matter for prax's use, where only json is sent.

**Review by turbocrime — COMMENTED** [review]
presently a draft, uploaded to show wip

File: packages/chrome-offscreen-worker/src/control.ts [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+      const sessionPort = chrome.runtime.connect({ name: offscreenHash });
+      sessionPort.onDisconnect.addListener(() => {
+        this.workers.forEach(p => p.disconnect());
+        this.workers.clear();
+      });
+      resolveConnection(sessionPort);
+    }
+
+    return offscreenConnection;
+  }
```
Comment (turbocrime) [inline packages/chrome-offscreen-worker/src/control.ts:112]: this method will be refactored, as it presently tries to begin using the document too quickly.

File: packages/chrome-offscreen-worker/src/worker.ts (hunk 9 above)
Comment (turbocrime) [reply packages/chrome-offscreen-worker/src/worker.ts]: only message events should go into the worker. the other types error and messageerror may also be emitted from the worker.

**Comment by turbocrime** [comment]
this is pretty close, but i had other tasks today and now i'm out of time. tl;dr:

a transaction build can take place successfully, but it seems like sessions live a bit too long.

workers should be redesigned to terminate themselves when possible.

be aware that when the `createDocument` method returns, offscreen is not ready.

this does not yet support use from multiple scripts - different global contexts will track offscreen use independently, and kill the shared document early.

it might be good to check if terminating the offscreen document actually kills the web workers, or if it's cleaned up too fast to issue termination commands.

this tool when complete would be incredibly useful to extension developers in general, so i'm likely to publish this polyfill independently to NPM.

**Comment by TalDerei** [comment]
`chrome-offscreen-worker` package in this PR will be replaced with https://github.com/turbocrime/chrome-offscreen-worker/tree/main/package

**Comment by TalDerei** [comment]
advocating to close the intermediate polyfill step in favor of proceeding _directly_ with the migration.

**Pull request closed without merging**